lib/attrsets.nix

```
@@ -1208,6 +1208,98 @@ rec {
     in
     recurse [ ] set;
 
+  /**
+    Recursively apply a function to each leaf attribute of the potentially
+    nested attribute set and combine the results in a list.
+
+    The first argument to the mapping function is a list of the attribute names
+    that form the path to the leaf attribute. The second argument will never
+    be an attribute set.
+
+    # Inputs
+
+    `f`
+
+    : A function, given an attribute’s path and value, returns a new value.
+
+    `attrs`
+
+    : Attribute set to map over.
+
+    # Type
+
+    ```
+    mapAttrsToListRecursive :: ([String] -> a -> b) -> AttrSet -> [b]
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.mapAttrsToListRecursive` usage example
+
+    ```nix
+    mapAttrsRecursive (path: value: "${concatStringsSep "." path}=${value}")
+      { n = { a = "A"; m = { b = "B"; c = "C"; }; }; d = "D"; }
+    => [ "n.a=A" "n.m.b=B" "n.m.c=C" "d=D" ]
+    ```
+    :::
+  */
+  mapAttrsToListRecursive =
+    f:
+    set:
+    mapAttrsToListRecursiveCond (as: true) f set;
+
+  /**
+    Like `mapAttrsToListRecursive`, but it takes an additional predicate to
+    decide whether to recurse into an attribute set. If the predicate returns
+    false, `mapAttrsToListRecursiveCond` does not recurse, but instead applies
+    the mapping function. If the predicate returns true, it does recurse
+    without applying the mapping function.
+
+    # Inputs
+
+    `pred`
+
+    : Predicate to decide whether to recurse into an attribute set.
+
+    `f`
+
+    : A function, given an attribute’s path and value, returns a new value.
+
+    `attrs`
+
+    : Attribute set to map over.
+
+    # Type
+    ```
+    mapAttrsToListRecursiveCond :: (AttrSet -> Bool) -> ([String] -> a -> b) -> AttrSet -> AttrSet
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.mapAttrsToListRecursiveCond` usage example
+
+    Map potentially nested set of derivations to a list of their `name` attribute.
+    ```nix
+    mapAttrsRecursiveCond
+      (as: !(lib.isDerivation as))
+      (x: x.name)
+      attrs
+    ```
+    :::
+  */
+  mapAttrsToListRecursiveCond =
+    cond:
+    f:
+    set:
+    let
+      recurse = path:
+        mapAttrsToList
+          (name: value:
+            if isAttrs value && cond value
```

could also take the current path and/or the name.
```suggestion
            if isAttrs value && cond (path ++ [ name ]) value
```

---

That’s a good idea. Applied as suggested.